### ISSUE TITLE
Rerunning checkout_externals without `-o` when there are already optional components checked out should update those optional components

### ISSUE THREAD

**Issue by billsacks** [ESMCI/manage_externals#155]
### Summary of Issue:

If you initially run `checkout_externals -o` to check out optional components (or run `checkout_externals EXTERNAL_NAME` on some specific optional external), then later rerun `checkout_externals` without the optional flag, any existing optional components are not updated. I feel that this is a bug, and an important one at that.

An important requirement that guided the design of manage_externals from the beginning is stated in the README file:

```
Running checkout_externals without the '--status' option will always attempt to
synchronize the working copy to exactly match the externals description.
```

However, the current behavior violates that requirement. Of course, it's reasonable that an optional component may not exist at all. But I feel that it's important that, if an optional component does exist, that it should always be checked out at the appropriate hash / tag after running `checkout_externals`. I feel that the current behavior is too error-prone: it's too easy for a user to forget the `-o` flag when updating their sandbox, and then running a case with the wrong externals. (When they do their initial checkout, it will be obvious if they needed the `-o` flag because they'll get a failure when creating the case if they try to do so with a missing external that is needed for their case. But this omission will often *not* be obvious when they are updating their externals.)

I haven't looked at the implementation, but I imagine this change will involve something like changing the behavior so that an already-existing external is treated as required, regardless of whether the Externals file lists it as optional.

### Expected behavior and actual behavior:

(Described above.)

### Steps to reproduce the problem (should include model description file(s) or link to public repository):

- Clone the CESM repository

- Run `manage_externals/checkout_externals -o`

- cd to `components/mom` and run `git checkout mi_20200820`

- Back at the top level, run `manage_externals/checkout_externals` without the `-o` flag

- Run `manage_externals/checkout_externals -S -v` and observe that MOM has not been re-synced

### What is the changeset ID of the code, and the machine you are using:

Whatever is included in the current CESM master.

### have you modified the code? If so, it must be committed and available for testing:

No

### Screen output or log file showing the error message and context:

N/A

----

@jedwards4b @gold2718 @alperaltuntas @mvertens

**Comment by jtruesdal** [comment]
@gold2718 @MiCurry Thanks for the heads up we will make sure to let the MPAS developers know that in addition to the initial MPAS check out they will have to continue to use the -o option when they wish to update the checked-out code.

**Comment by johnpaulalex** [comment]
Slight update for reproducing: checking out mi_20200820 leads to other trouble when running checkout_externals:
...
Processing externals description file : ../Externals_MOM.cfg
Processing submodules description file : /home/user/CESM/components/mom/MOM6/.gitmodules
ERROR:root:Cannot checkout "pkg/mom6_da_hooks" as a submodule, repo not found in .gitmodules file
...

But checking out a more recent tag works fine (I used mi_20210709).

**Comment by fischer-ncar** [comment]
@johnpaulalex I was able to checkout mi_20200820 and get pkg/mom6_da_hooks.  Are you able to reproduce this error, and what steps did you take to get the error?

**Comment by johnpaulalex** [comment]
Hi Chris, here's what I did to reproduce -- note I'm checking out a separate copy of manage_externals because I want to work on that code (to fix this issue). Maybe the CESM tag I'm using is bad somehow?

git clone -b cesm2_3_beta09 https://github.com/ESCOMP/CESM 
git clone -b master https://github.com/ESMCI/manage_externals myme
cd CESM
../myme/checkout_externals -o
cd components/mom && git checkout mi_20200820
../myme/checkout_externals   <-- mom6_da_hooks error happens here
cd components/mom && git checkout mi_20210709
../myme/checkout_externals  <-- works fine

**Comment by fischer-ncar** [comment]
@johnpaulalex The cesm2_3_beta09 tag is fine.  The issue is mi_20200820 has different submodules than mi_220624 and mi_20210709.  This is causing manage externals to trip up.  A couple of work arounds is to use cesm2_3_alpha02c or remove component/mom/MOM6.  Do you have a specific reason for using an older MOM6 tag?

**Comment by johnpaulalex** [comment]
Nope, just wanted to reproduce the bug (and make sure I wasn't doing something else wrong by working around it). I'll use mi_20210709 for now.

**Comment by billsacks** [comment]
Resolved by #172 - thank you @johnpaulalex !
